### PR TITLE
Series

### PR DESCRIPTION
**Pull request by Joydev**
Series playlist

### CODE REVIEW

File: frontend/src/public/series.11tydata.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }) => ({
+      title,
+      dateForHumans,
+      dateUncertain,
+      location,
+      locationUncertain,
+      category,
+      languages,
+      duration,
+    }),
```
Comment (anantakrishna) [inline frontend/src/public/series.11tydata.js:31]: This block seems irrelevant to the series data.

**Review by anantakrishna — CHANGES_REQUESTED** [review]
Let's finish the partials. I see that you have not used them anywhere. Without using, you cannot understand whether you implemented them properly.

Use the partials in the home page and single audio page.

File: frontend/src/_includes/partials/audio-actions.html [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        d="M8 10h.01M12 10h.01M16 10h.01M9 16H5a2 2 0 01-2-2V6a2 2 0 012-2h14a2 2 0 012 2v8a2 2 0 01-2 2h-5l-5 5v-5z"
+      />
+    </svg>
+    <span class="hidden md:inline">Feedback</span>
+  </a>
+  <!-- Share -->
+  <a
+    href="/audios/{{ audioObj.fileId }}/"
+    class="inline-flex cursor-pointer items-center space-x-1"
+    x-data="webshare('{{ audioObj.title | replace: "'", "\\'" | escape }}', '/audios/{{ audioObj.fileId }}/')"
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/audio-actions.html:49]: Make both `fileId` and `title` separate parameters of this partial. Don't pass `audioObj` as a whole.

File: frontend/src/_includes/partials/audio-duration.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  viewBox="0 0 24 24"
+  class="mx-1 my-2 w-4 stroke-current"
+  stroke-width="0.2"
+>
+  <path
+    fill="currentColor"
+    d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8s8 3.58 8 8s-3.58 8-8 8zm-.22-13h-.06c-.4 0-.72.32-.72.72v4.72c0 .35.18.68.49.86l4.15 2.49c.34.2.78.1.98-.24a.71.71 0 0 0-.25-.99l-3.87-2.3V7.72c0-.4-.32-.72-.72-.72z"
+  />
+</svg>
+<p>{{ audioDuration | duration }}</p>
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/audio-duration.html:12]: Theoretically, this partial can be used not only for audios. Therefore remove any mention of `audio` from the partial name and the parameter.

File: frontend/src/_includes/partials/audio-tags.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<div title="Unique file identifier">#{{ audioObj.fileId }}</div>
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/audio-tags.html:1]: Don't include `obj` in the parameter name. It's ugly and redundant.

Do it everywhere.

File: frontend/src/_includes/partials/colored-badges.html [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      style="--percentage: {% if audioObj.percentage > 0 -%}{{ audioObj.percentage | percentage }}{%- else -%}100%{%- endif -%}"
+    ></div>
+    <div class="w-full whitespace-nowrap px-1.5 pb-[1px] opacity-95">
+      {%- if audioObj.percentage > 0 -%}
+        Srila Gurudeva {{ audioObj.percentage | percentage }}
+      {%- else -%}
+        Other guru-varga
+      {%- endif -%}
+    </div>
+  </div>
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/colored-badges.html:24]: Let's split this into two partials, each will get only required value, not the whole audio object.

File: frontend/src/_includes/partials/audio-tags.html (hunk 4 above)
Comment (anantakrishna) [reply frontend/src/_includes/partials/audio-tags.html:1]: Also, move the surrounding div here as well, it's a non-separable part of the tags partial.

**Review by anantakrishna — CHANGES_REQUESTED** [review]

File: frontend/eleventy.config.cjs [hunk 6]
```diff
@@ -71,6 +71,17 @@ module.exports = function (eleventyConfig) {
     formatReducedPrecisionDate
   );
 
+  //audio actions filter -Download
+  eleventyConfig.addFilter(
+    'audio_action_download',
```
Comment (anantakrishna) [inline frontend/eleventy.config.cjs:76]: Let's call it `toFileURL`.

File: frontend/eleventy.config.cjs [hunk 7]
```diff
@@ -71,6 +71,17 @@ module.exports = function (eleventyConfig) {
...
+    (downloadFile) => `${process.env.STORAGE_BASE_URL}/` + downloadFile + `.mp3`
```
Comment (anantakrishna) [inline frontend/eleventy.config.cjs:77]: The semantics of the input parameter is not `downloadFile`. It's `fileId`.

File: frontend/eleventy.config.cjs [hunk 8]
```diff
@@ -71,6 +71,17 @@ module.exports = function (eleventyConfig) {
...
+  );
+
+  eleventyConfig.addFilter(
+    'audio_action_feedback',
```
Comment (anantakrishna) [inline frontend/eleventy.config.cjs:81]: Rename to `toFeedbackURL`.

File: frontend/eleventy.config.cjs [hunk 9]
```diff
@@ -71,6 +71,17 @@ module.exports = function (eleventyConfig) {
...
+    (feedbackurl) => `${process.env.FEEDBACK_FORM_AUDIOS}` + feedbackurl
```
Comment (anantakrishna) [inline frontend/eleventy.config.cjs:82]: Again, the parameter name does not reflect its semantics.

File: frontend/src/_data/series.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+const { readItems } = require('@directus/sdk');
+
+/**
+ * @param {EleventyGlobalData} data
+ */
+
+module.exports = ({ directus }) =>
+  directus.request(
+    readItems('series', {
+      fields: ['*.*.*'],
```
Comment (anantakrishna) [inline frontend/src/_data/series.js:14]: We don't need all that data. Pick only necessary fields.

File: frontend/src/_data/series.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+      filter: { status: { _eq: 'published' } },
+      // Also, we assume that the `QUERY_LIMIT_MAX` env variable is not set on the server.
+      // Fetching only few items in the development ymode to speed up rebuilding the pages
```
Comment (anantakrishna) [inline frontend/src/_data/series.js:17]: This line regarding the development mode is not relevant anymore.

File: frontend/src/_includes/partials/actions.html [hunk 12]
```diff
@@ -0,0 +1,72 @@
+<div class="mb-1 flex space-x-1">
+  <!-- Download -->
+  <a
+    href="{{ audioFileId | audio_action_download }}"
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/actions.html:4]: Don't pollute the names of the identifiers with redundant specifiers. The context of this template is one single audio. The parameter name semantically is `fileId`.

Also, now I see that it may be proper to pass the whole record to the partial, and let it use whatever properties are needed. Otherwise the calling side needs to know which properties to pass. This breaks the principle of the separation of concerns.

By the way, in this case it's better to include the `audio` term in the partial's file name. Because we can have actions for other type of content.

File: frontend/src/_includes/partials/audio-tags.html [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+{% if audioTagsParam.dateForHumans -%}
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/audio-tags.html:2]: What a terrible `audioTagsParam`. See my other comments related to naming.

File: frontend/src/audio.html [hunk 14]
```diff
@@ -1,189 +1,33 @@
...
-      >
-        <path
-          fill="currentColor"
-          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8s8 3.58 8 8s-3.58 8-8 8zm-.22-13h-.06c-.4 0-.72.32-.72.72v4.72c0 .35.18.68.49.86l4.15 2.49c.34.2.78.1.98-.24a.71.71 0 0 0-.25-.99l-3.87-2.3V7.72c0-.4-.32-.72-.72-.72z"
-        />
-      </svg>
-      <p class="pt-1">{{ audio.duration | duration }}</p>
-    </span>
+    <div class="inline-flex content-end items-center">
+      {% render 'partials/duration.html', durationParam: audio.duration -%}
```
Comment (anantakrishna) [inline frontend/src/audio.html:13]: Here and everywhere you can use the special syntax for passing a single parameter: https://liquidjs.com/tags/render.html#The-with-Parameter.

And don't include `param` into the parameter name! Everywhere.

File: frontend/src/audio.html [hunk 15]
```diff
@@ -1,189 +1,33 @@
...
-          <path
-            stroke-linecap="round"
-            stroke-linejoin="round"
-            d="M7.217 10.907a2.25 2.25 0 100 2.186m0-2.186c.18.324.283.696.283 1.093s-.103.77-.283 1.093m0-2.186l9.566-5.314m-9.566 7.5l9.566 5.314m0 0a2.25 2.25 0 103.935 2.186 2.25 2.25 0 00-3.935-2.186zm0-12.814a2.25 2.25 0 103.933-2.185 2.25 2.25 0 00-3.933 2.185z"
-          />
-        </svg>
-        <span class="hidden md:inline">Share</span>
-      </a>
-    </div>
+    {% render 'partials/actions.html', audioFileId: audio.fileId, audioTitle: audio.title -%}
```
Comment (anantakrishna) [inline frontend/src/audio.html:30]: Check if we can remove the calculated `fileURL` property of the audio data.

**Review by anantakrishna — CHANGES_REQUESTED** [review]

File: frontend/eleventy.config.cjs (hunk 6 above)
Comment (Joydev) [reply frontend/eleventy.config.cjs:76]: done

File: frontend/eleventy.config.cjs (hunk 7 above)
Comment (Joydev) [reply frontend/eleventy.config.cjs:77]: changed to `fileId`

File: frontend/eleventy.config.cjs (hunk 8 above)
Comment (Joydev) [reply frontend/eleventy.config.cjs:81]: done

File: frontend/eleventy.config.cjs (hunk 9 above)
Comment (Joydev) [reply frontend/eleventy.config.cjs:82]: changed to `fileId`

File: frontend/src/_data/series.js (hunk 11 above)
Comment (Joydev) [reply frontend/src/_data/series.js:17]: removed

File: frontend/src/_includes/partials/actions.html (hunk 12 above)
Comment (Joydev) [reply frontend/src/_includes/partials/actions.html:4]: we can't pass whole record here as fileId specifier is different for series page (`fileId: audio.audios_id.id`) and audio page (`fileId: audio.fileId`)

File: frontend/src/_includes/partials/audio-tags.html (hunk 4 above)
Comment (Joydev) [reply frontend/src/_includes/partials/audio-tags.html:1]: done

File: frontend/src/_includes/partials/audio-tags.html (hunk 13 above)
Comment (Joydev) [reply frontend/src/_includes/partials/audio-tags.html:2]: done

File: frontend/src/_includes/partials/audio-actions.html (hunk 2 above)
Comment (Joydev) [reply frontend/src/_includes/partials/audio-actions.html]: done

File: frontend/src/audio.html (hunk 14 above)
Comment (Joydev) [reply frontend/src/audio.html:13]: done

File: frontend/eleventy.config.cjs [hunk 16]
```diff
@@ -71,6 +71,17 @@ module.exports = function (eleventyConfig) {
...
 
+  //audio actions filter -Download
+  eleventyConfig.addFilter(
+    'toFileURL',
+    (fileId) => `${process.env.STORAGE_BASE_URL}/` + fileId + `.mp3`
+  );
+
+  eleventyConfig.addFilter(
+    'toFeedbackURL',
+    (fileId) => `${process.env.FEEDBACK_FORM_AUDIOS}` + fileId
```
Comment (anantakrishna) [inline frontend/eleventy.config.cjs:82]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#string_interpolation

File: frontend/src/_includes/partials/duration.html [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+  viewBox="0 0 24 24"
+  class="mx-1 my-2 w-4 stroke-current"
+  stroke-width="0.2"
+>
+  <path
+    fill="currentColor"
+    d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8s8 3.58 8 8s-3.58 8-8 8zm-.22-13h-.06c-.4 0-.72.32-.72.72v4.72c0 .35.18.68.49.86l4.15 2.49c.34.2.78.1.98-.24a.71.71 0 0 0-.25-.99l-3.87-2.3V7.72c0-.4-.32-.72-.72-.72z"
+  />
+</svg>
+<p>{{ duration | duration }}</p>
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/duration.html:12]: Let's rename the `duration` filter to more descriptive `formatDurationForHumans`

File: frontend/src/_includes/partials/percentage.html [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    {%- endif -%}
+  "
+>
+  <div
+    class="absolute z-auto h-full w-[var(--percentage)] bg-gray-300 shadow-none"
+    style="--percentage: {% if percentage > 0 -%}{{ percentage | percentage }}{%- else -%}100%{%- endif -%}"
+  ></div>
+  <div class="w-full whitespace-nowrap px-1.5 pb-[1px] opacity-95">
+    {%- if percentage > 0 -%}
+      Srila Gurudeva {{ percentage | percentage }}
```
Comment (anantakrishna) [inline frontend/src/_includes/partials/percentage.html:18]: Here also, the filter name is not descriptive enough. Let's rename it to `formatFractionAsPercentage`.

File: frontend/src/audio.html [hunk 19]
```diff
@@ -1,189 +1,37 @@
...
-            {%- endif %}
-          </div>
-        {%- endif %}
-        {% if audio.category -%}
-          <div>{{ audio.category }}</div>
-        {%- endif %}
-        {% for language in audio.languages -%}
-          <div>{{ language }}</div>
-        {%- endfor %}
+        {% render 'partials/audio-tags.html', audio: audio -%}
```
Comment (anantakrishna) [inline frontend/src/audio.html:23]: Use the `with` expression.

File: frontend/src/serie.html [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+            </div>
+          </div>
+          <!-- Second line -->
+          <div class="mt-1 flex items-start justify-between space-x-2 pr-2 text-xs">
+            <div class="flex flex-grow flex-wrap items-start justify-between">
+              <!-- Tags -->
+              <div class="mb-1 flex flex-wrap content-between items-baseline [&>*]:mb-1 [&>*]:mr-1 [&>*]:border-r [&>*]:border-gray-300 [&>*]:pr-1 last:[&>*]:border-0">
+                <div title="Unique file identifier">#{{ audio.audios_id.id }}</div>
+                {% render 'partials/audio-tags.html', audio: audio.audios_id -%}
+              </div>
```
Comment (anantakrishna) [inline frontend/src/serie.html:47]: Move the surrounding div into the `tags` partial, as well as the file ID.

File: frontend/src/serie.html [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+              <div class="mb-1 flex items-center space-x-1">
+                <!-- Colored badges -->
+                {% assign audioPercentage = audio.audios_id.percentage %}
+                {% render 'partials/percentage.html' with audioPercentage as percentage -%}
+                {% assign audiosoundQualityRating = audio.audios_id.soundQualityRating %}
+                {% render 'partials/sound-quality.html' with audiosoundQualityRating as soundQualityRating -%}
+              </div>
+            </div>
+            <!-- Actions -->
+            {% render 'partials/audio-actions.html', fileId: audio.audios_id.id, title: audio.audios_id.title -%}
```
Comment (anantakrishna) [inline frontend/src/serie.html:57]: After we moved all the HTML into a partial, there is no need for the comment above it. It's obvious what follows without a comment.

**Review by anantakrishna — CHANGES_REQUESTED** [review]

File: frontend/src/_includes/partials/actions.html (hunk 12 above)
Comment (anantakrishna) [reply frontend/src/_includes/partials/actions.html]: Can't we pass `audio` in one case and `audio.audio_id` in another?

File: frontend/src/audio.html (hunk 15 above)
Comment (anantakrishna) [reply frontend/src/audio.html]: Have you checked?

File: frontend/src/audio.html [hunk 22]
```diff
@@ -1,189 +1,27 @@
...
-        <path
-          fill="currentColor"
-          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8s8 3.58 8 8s-3.58 8-8 8zm-.22-13h-.06c-.4 0-.72.32-.72.72v4.72c0 .35.18.68.49.86l4.15 2.49c.34.2.78.1.98-.24a.71.71 0 0 0-.25-.99l-3.87-2.3V7.72c0-.4-.32-.72-.72-.72z"
-        />
-      </svg>
-      <p class="pt-1">{{ audio.duration | duration }}</p>
-    </span>
+    <div class="inline-flex content-end items-center">
+      {% render 'partials/duration.html' with audio.duration as duration -%}
+    </div>
```
Comment (anantakrishna) [inline frontend/src/audio.html:13]: This surrounding `div` is an inherent part of the duration block. Why not to move it inside the partial?

File: frontend/src/audio.html [hunk 23]
```diff
@@ -1,189 +1,27 @@
...
-          <path
-            stroke-linecap="round"
-            stroke-linejoin="round"
-            d="M7.217 10.907a2.25 2.25 0 100 2.186m0-2.186c.18.324.283.696.283 1.093s-.103.77-.283 1.093m0-2.186l9.566-5.314m-9.566 7.5l9.566 5.314m0 0a2.25 2.25 0 103.935 2.186 2.25 2.25 0 00-3.935-2.186zm0-12.814a2.25 2.25 0 103.933-2.185 2.25 2.25 0 00-3.933 2.185z"
-          />
-        </svg>
-        <span class="hidden md:inline">Share</span>
-      </a>
-    </div>
+    {% render 'partials/audio-actions.html', fileId: audio.fileId, title: audio.title -%}
```
Comment (anantakrishna) [inline frontend/src/audio.html:24]: I think here we can also pass the whole `audio` object to the partial, and let it decide what it needs.

**Review by anantakrishna — CHANGES_REQUESTED** [review]

File: frontend/src/audio.html (hunk 22 above)
Comment (Joydev) [reply frontend/src/audio.html]: done

File: frontend/src/audio.html (hunk 23 above)
Comment (Joydev) [reply frontend/src/audio.html]: done
